File: week-01/homework/lc560_subarray_sum_equals_k/lc560_subarray_sum_equals_k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        nums_sum_dict = {}
        nums_sum = count = 0
        # i <- 1..n, j <- 1..i，几个s[i] - s[j] = k
        for num in nums:
            nums_sum += num

            # 当s[j] = 0, s[i] = k有且仅有一个
            if nums_sum == k:
                count += 1
            # 当s[j] > 0, 有几个s[j]满足s[j] = s[i] - k
            if nums_sum - k in nums_sum_dict:
                count += nums_sum_dict[nums_sum-k]

            # 后置存储nums_sum的个数，避免k=0的情况
            if nums_sum in nums_sum_dict:
                nums_sum_dict[nums_sum] += 1
            else:
                nums_sum_dict[nums_sum] = 1
        return count

File: week-01/homework/lc560_subarray_sum_equals_k/lc560_subarray_sum_equals_k.scala
object Solution {
    def subarraySum(nums: Array[Int], k: Int): Int = {
        var nums_sum = 0
        var count = 0
        var nums_sum_array = Array[Int]()
        nums.map(x => {
            nums_sum = nums_sum + x
            nums_sum_array = nums_sum_array :+ nums_sum
        })

        // 先求出直接等于k的nums_sum个数
        count = nums_sum_array.filter(x => x == k).length


        for(j <- Range(0, nums_sum_array.length)) {
            // upper bound
            for (i <- (j + 1) until nums_sum_array.length) {
                // 再找出所有满足s[i] - s[j] = k的个数 (j > 0)
                if (nums_sum_array(i) - nums_sum_array(j) == k) {
                    count = count + 1
                }
            }
        }
        count
    }
}